# importing libraries

In [1]:
#importing
import pandas as pd
import numpy  as np
import  os
import matplotlib.pyplot as  plt
import seaborn as sns
import  scipy

In [2]:
customer_data = pd.read_csv(r'C:\Users\makay\Documents\Pythons files\Master folder\02 Data\Original data\customers 2.csv')

In [3]:
customer_data.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

# Data Wrangling

In [4]:
#renaming  columns
customer_data.rename(columns = {'First Name':'First_name','Surnam':'Last_name','STATE':'State','n_dependants':'number_of_dependents','fam_status':'Family_status'}, inplace = True)

In [5]:
customer_data.head()

user_id First_name Last_name  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependents Family_status  income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

In [6]:
customer_data.dtypes

user_id                  int64
First_name              object
Last_name               object
Gender                  object
State                   object
Age                      int64
date_joined             object
number_of_dependents     int64
Family_status           object
income                   int64
dtype: object

In [7]:
#Checking for  missing values
customer_data.isnull().sum()

user_id                     0
First_name              11259
Last_name                   0
Gender                      0
State                       0
Age                         0
date_joined                 0
number_of_dependents        0
Family_status               0
income                      0
dtype: int64

# first  name is the only column to have any missing values

In [9]:
#subsetting with a  filter on the first name column
customer_data_clean = customer_data[customer_data['First_name'].isnull() == False]

In [10]:
customer_data.shape

(206209, 10)

In [11]:
customer_data_clean.shape

(194950, 10)

# By making the first name column filter with false it took out the 11259 rows leaving me with only clean data

In [12]:
#Checking for mixed data types
for col in customer_data_clean.columns.tolist():
    weird = (customer_data_clean[[col]].applymap(type) != customer_data_clean[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (customer_data_clean[weird]) > 0:
        print (col)

#  no mixed data types

In [13]:
#Checking for duplicates 
customer_data_duplicates = customer_data_clean[customer_data_clean.duplicated()]

In [14]:
customer_data_duplicates

Empty DataFrame
Columns: [user_id, First_name, Last_name, Gender, State, Age, date_joined, number_of_dependents, Family_status, income]
Index: []

# no duplicates found

In [15]:
#importing instacart data to combine with customer dat
ords_prods_merged = pd.read_pickle(r'C:\Users\makay\Documents\Pythons files\Master folder\02 Data\Prepared data\4.8 ords_prods_mergedd.pkl')

In [16]:
ords_prods_merged.head()

Unnamed: 0_x  order_id  user_id eval_set  order_number  order_dow  \
0             0   2539329        1    prior             1          2   
1             1   2398795        1    prior             2          3   
2             2    473747        1    prior             3          3   
3             3   2254736        1    prior             4          4   
4             4    431534        1    prior             5          4   

   order_hour_of_day  days_since_prior_order  first_time_ordering  product_id  \
0                  8                     NaN                 True       196.0   
1                  7                    15.0                False       196.0   
2                 12                    21.0                False       196.0   
3                  7                    29.0                False       196.0   
4                 15                    28.0                False       196.0   

   ...    price_range_loc      busiet day         busiest_days  \
0  ...  Mid-range product  Regularly busy  Regularly busy days   
1  ...  Mid-range product  Regularly busy         slowest days   
2  ...  Mid-range product  Regularly busy         slowest days   
3  ...  Mid-range product      Least busy  Regularly busy days   
4  ...  Mid-range product      Least busy  Regularly busy days   

  busiest_period_of_day  max_order  loyalty_flag  Average_price  \
0        Average orders         10  New customer           14.0   
1        Average orders         10  New customer           14.0   
2        Average orders         10  New customer           14.0   
3        Average orders         10  New customer           14.0   
4        Average orders         10  New customer           14.0   

   Spending_flag Customer_frequency order_frequency_flag  
0  High  Spender               30.0     Regular customer  
1  High  Spender               30.0     Regular customer  
2  High  Spender               30.0     Regular customer  
3  High  Spender               30.0     Regular customer  
4  High  Spender               30.0     Regular customer  

[5 rows x 28 columns]

In [17]:
customer_data_clean.head()

user_id First_name Last_name  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependents Family_status  income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

In [18]:
ords_prods_merged = ords_prods_merged.drop('_merge', 1)

C:\Users\makay\AppData\Local\Temp\ipykernel_10324\1233194729.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  ords_prods_merged = ords_prods_merged.drop('_merge', 1)


In [19]:
ords_prods_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32434212 entries, 0 to 32434211
Data columns (total 27 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0_x            int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   eval_set                object 
 4   order_number            int64  
 5   order_dow               int64  
 6   order_hour_of_day       int64  
 7   days_since_prior_order  float64
 8   first_time_ordering     bool   
 9   product_id              float64
 10  add_to_cart_order       float64
 11  reordered               float64
 12  Unnamed: 0_y            int64  
 13  product_name            object 
 14  aisle_id                int64  
 15  department_id           int64  
 16  prices                  float64
 17  price_range_loc         object 
 18  busiet day              object 
 19  busiest_days            object 
 20  busiest_period_of_day   object 
 21  max_order               int64

memory size is too lalrge so ahving to change data types to cut back on size

In [20]:
ords_prods_merged['Customer_frequency'] = ords_prods_merged['Customer_frequency'].astype('float16')

In [21]:
ords_prods_merged['Average_price'] = ords_prods_merged['Average_price'].astype('float16')

In [22]:
ords_prods_merged['max_order'] = ords_prods_merged['max_order'].astype('int8')

In [23]:
ords_prods_merged['aisle_id'].max()

134

In [24]:
ords_prods_merged['aisle_id'] = ords_prods_merged['aisle_id'].astype('int8')

In [25]:
ords_prods_merged['prices'].max()

99999.0

In [26]:
ords_prods_merged['prices'] = ords_prods_merged['prices'].astype('float16')

In [28]:
ords_prods_merged = ords_prods_merged.drop('Unnamed: 0_x', 1)

C:\Users\makay\AppData\Local\Temp\ipykernel_10324\1881214624.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  ords_prods_merged = ords_prods_merged.drop('Unnamed: 0_x', 1)


In [29]:
ords_prods_merged = ords_prods_merged.drop('Unnamed: 0_y', 1)

C:\Users\makay\AppData\Local\Temp\ipykernel_10324\2711870176.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  ords_prods_merged = ords_prods_merged.drop('Unnamed: 0_y', 1)


In [30]:
ords_prods_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32434212 entries, 0 to 32434211
Data columns (total 25 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
 7   first_time_ordering     bool   
 8   product_id              float64
 9   add_to_cart_order       float64
 10  reordered               float64
 11  product_name            object 
 12  aisle_id                int8   
 13  department_id           int64  
 14  prices                  float16
 15  price_range_loc         object 
 16  busiet day              object 
 17  busiest_days            object 
 18  busiest_period_of_day   object 
 19  max_order               int8   
 20  loyalty_flag            object 
 21  Average_price           float

In [31]:
ords_prods_merged['add_to_cart_order'].max()

145.0

In [32]:
ords_prods_merged['add_to_cart_order'] = ords_prods_merged['add_to_cart_order'].astype('float16')

In [33]:
ords_prods_merged['product_name'] = ords_prods_merged['product_name'].astype('category')

In [34]:
ords_prods_merged['loyalty_flag'] = ords_prods_merged['loyalty_flag'].astype('category')

In [36]:
ords_prods_merged['Spending_flag'] = ords_prods_merged['Spending_flag'].astype('category')

In [37]:
ords_prods_merged['order_frequency_flag'] = ords_prods_merged['order_frequency_flag'].astype('category')

In [38]:
ords_prods_merged = ords_prods_merged.drop('eval_set', 1)

C:\Users\makay\AppData\Local\Temp\ipykernel_10324\1318987309.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  ords_prods_merged = ords_prods_merged.drop('eval_set', 1)


In [39]:
ords_prods_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32434212 entries, 0 to 32434211
Data columns (total 24 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   order_number            int64   
 3   order_dow               int64   
 4   order_hour_of_day       int64   
 5   days_since_prior_order  float64 
 6   first_time_ordering     bool    
 7   product_id              float64 
 8   add_to_cart_order       float16 
 9   reordered               float64 
 10  product_name            category
 11  aisle_id                int8    
 12  department_id           int64   
 13  prices                  float16 
 14  price_range_loc         object  
 15  busiet day              object  
 16  busiest_days            object  
 17  busiest_period_of_day   object  
 18  max_order               int8    
 19  loyalty_flag            category
 20  Average_price           float16 
 21  Spendi

In [40]:
ords_prods_merged['order_dow'] = ords_prods_merged['order_dow'].astype('int8')

In [41]:
ords_prods_merged['order_hour_of_day'] = ords_prods_merged['order_hour_of_day'].astype('int8')

In [42]:
ords_prods_merged['days_since_prior_order'] = ords_prods_merged['days_since_prior_order'].astype('float16')

In [43]:
ords_prods_merged['product_id'] = ords_prods_merged['product_id'].astype('float16')

In [44]:
ords_prods_merged['reordered'] = ords_prods_merged['reordered'].astype('float16')

In [45]:
ords_prods_merged['department_id'] = ords_prods_merged['department_id'].astype('int8')

In [46]:
ords_prods_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32434212 entries, 0 to 32434211
Data columns (total 24 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   order_number            int64   
 3   order_dow               int8    
 4   order_hour_of_day       int8    
 5   days_since_prior_order  float16 
 6   first_time_ordering     bool    
 7   product_id              float16 
 8   add_to_cart_order       float16 
 9   reordered               float16 
 10  product_name            category
 11  aisle_id                int8    
 12  department_id           int8    
 13  prices                  float16 
 14  price_range_loc         object  
 15  busiet day              object  
 16  busiest_days            object  
 17  busiest_period_of_day   object  
 18  max_order               int8    
 19  loyalty_flag            category
 20  Average_price           float16 
 21  Spendi

In [47]:
# Merging Instacart data with Customer Data
instacart_merge = ords_prods_merged.merge(customer_data_clean, on = 'user_id', indicator = True)

In [48]:
instacart_merge.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  first_time_ordering  product_id  add_to_cart_order  \
0                     NaN                 True       196.0                1.0   
1                    15.0                False       196.0                1.0   
2                    21.0                False       196.0                1.0   
3                    29.0                False       196.0                1.0   
4                    28.0                False       196.0                1.0   

   reordered  ... First_name  Last_name  Gender    State Age date_joined  \
0        0.0  ...      Linda     Nguyen  Female  Alabama  31   2/17/2019   
1        1.0  ...      Linda     Nguyen  Female  Alabama  31   2/17/2019   
2        1.0  ...      Linda     Nguyen  Female  Alabama  31   2/17/2019   
3        1.0  ...      Linda     Nguyen  Female  Alabama  31   2/17/2019   
4        1.0  ...      Linda     Nguyen  Female  Alabama  31   2/17/2019   

  number_of_dependents Family_status  income _merge  
0                    3       married   40423   both  
1                    3       married   40423   both  
2                    3       married   40423   both  
3                    3       married   40423   both  
4                    3       married   40423   both  

[5 rows x 34 columns]

In [49]:
instacart_merge['_merge'].value_counts(dropna = False)

both          30657347
left_only            0
right_only           0
Name: _merge, dtype: int64

# Exporting successful Merge

In [50]:
instacart_merge.to_pickle(r'C:\Users\makay\Documents\Pythons files\Master folder\02 Data\Prepared data\4.9 instacart merged.pkl')